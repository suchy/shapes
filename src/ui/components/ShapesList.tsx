import { Fragment } from 'react';
import styled from 'styled-components';

import { useShallowSelector } from '../hooks/use-shallow-selector';
import { shapesSelector } from '../state/selectors';

import { ShapesListTitle } from './ShapesListTitle';
import { Shape } from './Shape';

const List = styled.ul`
  display: flex;
  flex-wrap: wrap;
  margin: 0 -1.2rem;
`;

const Item = styled.li`
  width: calc(50% - 2.4rem);
  margin: 0 1.2rem 4rem;
  position: relative;

  @media (min-width: 400px) {
    width: calc((100% / 3) - 2.4rem);
  }

  @media (min-width: 640px) {
    width: calc(25% - 2.4rem);
  }
`;

const ItemContainer = styled.div`
  width: 100%;
  padding-top: 100%;
  border: solid 0.1rem #f2f2f2;
  border-radius: 0.4rem;
  position: relative;
`;

const ItemShape = styled(Shape)`
  position: absolute;
  top: 18%;
  right: 18%;
  bottom: 18%;
  left: 18%;
`;

const EmptyMessage = styled.p``;

export const ShapesList = () => {
  const shapes = useShallowSelector(shapesSelector);

  if (!shapes.length) {
    return <EmptyMessage>No shapes matching criteria were found.</EmptyMessage>;
  }

  return (
    <Fragment>
      <ShapesListTitle />

      <List data-testid="ShapesList">
        {shapes.map((shape) => (
          <Item key={shape.id}>
            <ItemContainer>
              <ItemShape shape={shape} />
            </ItemContainer>
          </Item>
        ))}
      </List>
    </Fragment>
  );
};
