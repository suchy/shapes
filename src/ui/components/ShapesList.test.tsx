import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createReducer, configureStore } from '@reduxjs/toolkit';

import { ShapesList } from './ShapesList';

describe('ShapesList', () => {
  let store: any;

  describe('with shapes', () => {
    beforeEach(() => {
      store = configureStore({
        reducer: createReducer(
          {
            shapes: [{ id: 1, color: 'purple', shape: 'square' }],
            colorFilters: ['purple'],
            shapeFilters: ['square']
          },
          jest.fn()
        )
      });
    });

    it('should show shapes list', () => {
      const { getByTestId } = render(
        <Provider store={store}>
          <ShapesList />
        </Provider>
      );

      expect(getByTestId('ShapesList')).toBeVisible();
    });
  });

  describe('without shapes', () => {
    beforeEach(() => {
      store = configureStore({
        reducer: createReducer({ shapes: [] }, jest.fn())
      });
    });

    it('should match snapshot', () => {
      const { container } = render(
        <Provider store={store}>
          <ShapesList />
        </Provider>
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should show error message', () => {
      const { getByText } = render(
        <Provider store={store}>
          <ShapesList />
        </Provider>
      );

      expect(
        getByText('No shapes matching criteria were found.')
      ).toBeVisible();
    });
  });
});
