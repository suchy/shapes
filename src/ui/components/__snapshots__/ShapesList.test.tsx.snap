// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShapesList without shapes should match snapshot 1`] = `
<p
  class="sc-kEjbxe"
>
  No shapes matching criteria were found.
</p>
`;
